atc/wrappa/concurrent_request_limits_wrappa.go

```diff
@@ -0,0 +1,48 @@
+package wrappa
+
+import (
+	"net/http"
+
+	"code.cloudfoundry.org/lager"
+
+	"github.com/tedsuo/rata"
+)
+
+type ConcurrencyLimitsWrappa struct {
+	logger                  lager.Logger
+	concurrentRequestPolicy ConcurrentRequestPolicy
+}
+
+func NewConcurrencyLimitsWrappa(
+	logger lager.Logger,
+	concurrentRequestPolicy ConcurrentRequestPolicy,
+) Wrappa {
+	return ConcurrencyLimitsWrappa{
+		logger:                  logger,
+		concurrentRequestPolicy: concurrentRequestPolicy,
+	}
+}
+
+func (wrappa ConcurrencyLimitsWrappa) Wrap(handlers rata.Handlers) rata.Handlers {
+	wrapped := rata.Handlers{}
+
+	for name, handler := range handlers {
+		if wrappa.concurrentRequestPolicy.IsLimited(name) {
+			wrapped[name] = wrapHandler(
+				wrappa.logger,
+				wrappa.concurrentRequestPolicy.MaxConcurrentRequests(name),
+				handler,
+			)
+		} else {
+			wrapped[name] = handler
+		}
+	}
+
+	return wrapped
+}
+
+func wrapHandler(logger lager.Logger, limit int, handler http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusTooManyRequests)
+	})
```

Comment:
```suggestion
	handlerPool := semaphore.NewWeighted(int64(limit))
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if !handlerPool.TryAcquire(1) {
			w.WriteHeader(http.StatusTooManyRequests)
		} else {
			defer handlerPool.Release(1)
			handler.serveHTTP(w,r)
		}
	})
```
This is generally the right implementation, although we'll probably create a `Pool` interface to accomodate the metric feature. There should be a `HandlerPool(string action) Pool` method on `ConcurrentRequestPolicy` and then the pool could get passed in
```suggestion
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if !handlerPool.TryAcquire() {
			w.WriteHeader(http.StatusTooManyRequests)
		} else {
			defer handlerPool.Release()
			handler.serveHTTP(w,r)
		}
	})
```

Reply:
`semaphore` is `golang.org/x/sync/sempahore`.